### ISSUE TITLE
Break dependency cycles

### ISSUE THREAD

**Issue by vjeux** [postcss/postcss#1030]
I'm trying to use rollup to bundle postcss for prettier ( https://github.com/prettier/prettier/pull/1836 ) and unfortunately rollup doesn't handle cycles for commonjs modules properly. It is possible to use rollup but it involves some crazy hacks: https://github.com/styled-components/styled-components/pull/100

The cycles are due to the uses of inline require:
- https://github.com/postcss/postcss/blob/master/lib/container.es6#L581
- https://github.com/postcss/postcss/blob/master/lib/root.es6#L65-L66

I tried to fix them but it's not easy and it would require a non trivial refactor of the codebase.

It would make my life a lot easier as a consumer of postcss if sometimes in the future you could fix this! Thanks <3

**Comment by ai** [comment]
Hi. Yeap, I tried to avoid loop requires too, but failed in it :(.

So I am open for suggestions :).

**Comment by vjeux** [comment]
Usually the way to deal with those is to pass the parse function all the way down. This way when you are in container, you can call parse without require it.

**Comment by ai** [comment]
But `new Rule()` is a still valid API. And we can’t ask users to pass some extra argument.

But what do you think about this hack:

1. `Container` will accept extra option with `parse` function. In this case it will use this link.
2. If user call `new Container` without this extra option, `Container` will use `require()`.

Or Rollup anyway will raise a error, when it finds `require` in source?

**Comment by vjeux** [comment]
Oh, that's unfortunate.

The thing you could do is to make Rule module be a factory.

```js
module.exports = function(parser) {
  return class Rule { ... };
};
```

This is going to make the code look weird but would solve it.

I have no idea about rollup but it would still have a dependency so I guess it would break since it's analyzing this statically, but don't know, would have to check.

**Comment by ai** [comment]
It is not so easy to fix this code with factory, since we need not only `parse`, but also Container child classes 😕.

I know that other users sent PR to PostCSS for Rollup 
support. Maybe problem is not so big, and we get error only when somebody use that code with `require`? Can you check?

**Comment by vjeux** [comment]
I worked around in prettier so it's not a huge issue anymore. It would be nice if it works out of the box but oh well :)

**Comment by Jinjiang** [comment]
Sorry, I also meet this problem for browser env with Rollup. 🙁
Is it good enough to publish a "standalone" version together like babel or prettier in the coming versions? That would be more compatible IMO.
Thanks.

**Comment by ai** [comment]
@Jinjiang what you mean for “standalone” version? Prepack PostCSS into the single file by webpack?

**Comment by ai** [comment]
@Jinjiang I afraid that it will increase npm package size and will be used very very rare.

Can you use webpack to make that file in your system and then require this file by Rollup?

**Comment by Jinjiang** [comment]
No problem. I have done that with nothing error.

Btw the standalone package also could be published as an independent npm package to avoid increasing the size of the main package. And if there is an official one, maybe it will be widely used as an online repl, or in some code snippets/playground like codepen/jsfiddle/jsbin I guess. 😉

Thanks.

**Comment by ai** [comment]
> Btw the standalone package also could be published as an independent npm package to avoid increasing the size of the main package. And if there is an official one, maybe it will be widely used as an online repl, or in some code snippets/playground like codepen/jsfiddle/jsbin I guess. wink

Hm, it sounds good. If somebody will provide good scripts for it, I can release it.

**Comment by tivac** [comment]
Just in case this ends up helping someone else I wrote a little proof-of-concept[rollup plugin that uses webpack to bundle postcss](https://gist.github.com/tivac/d3dd0142e7c52d0d0a5a05b6686340e3).

This is a very silly thing to do, and I really wish it wasn't necessary. Seems to work ok though?

**Comment by ai** [comment]
The problem was fixed in 8.0 branch `ose`
